File: sdk/python/src/openlit/guard/sensitive_topic.py
"""
Sensitive topic detection guard.

Uses keyword/regex dictionaries for fast-path detection of sensitive
content categories.  An optional user-provided classifier handles
ambiguous cases.  Phases: preflight + postflight.
"""

from __future__ import annotations

import re
from typing import Callable, Dict, List, Optional, Set

from openlit.guard._base import Guard, GuardPhase, GuardResult

# Built-in category dictionaries.  Each value is a list of regex patterns
# that indicate the topic is present.
_DEFAULT_CATEGORIES: Dict[str, List[re.Pattern]] = {
    "violence": [
        re.compile(
            r"\b(?:kill|murder|assault|attack|weapon|bomb|shoot|stab|torture|terrorism|massacre|genocide)\b",
            re.IGNORECASE,
        ),
    ],
    "politics": [
        re.compile(
            r"\b(?:democrat|republican|election\s+fraud|political\s+party|vote\s+rigging|coup|insurrection)\b",
            re.IGNORECASE,
        ),
    ],
    "substance_use": [
        re.compile(
            r"\b(?:cocaine|heroin|methamphetamine|fentanyl|drug\s+(?:deal|traffick)|overdose|illegal\s+drugs)\b",
            re.IGNORECASE,
        ),
    ],
    "mental_health": [
        re.compile(
            r"\b(?:suicid(?:e|al)|self[- ]harm|eating\s+disorder|anorexia|bulimia)\b",
            re.IGNORECASE,
        ),
    ],
    "discrimination": [
        re.compile(
            r"\b(?:racial\s+slur|white\s+supremac|ethnic\s+cleansing|hate\s+(?:speech|crime))\b",
            re.IGNORECASE,
        ),
    ],
    "adult_content": [
        re.compile(
            r"\b(?:pornograph|explicit\s+sexual|nude\s+images|sex\s+trafficking)\b",
            re.IGNORECASE,
        ),
    ],
}


class SensitiveTopic(Guard):
    """
    Detects sensitive content categories.

    Parameters
    ----------
    action : str
        ``"warn"`` (default), ``"deny"``, or ``"redact"``.
    categories : set[str], optional
        Subset of built-in categories to enable.  ``None`` means all.
    custom_categories : dict, optional
        ``{category_name: [regex_string, ...]}`` to add.
    classifier : callable, optional
        ``Callable[[str], str | None]`` returning a category name or None.
    """

    name = "sensitive_topic"
    phases = (GuardPhase.PREFLIGHT, GuardPhase.POSTFLIGHT)

    def __init__(
        self,
        action: str = "warn",
        categories: Optional[Set[str]] = None,
        custom_categories: Optional[Dict[str, List[str]]] = None,
        classifier: Optional[Callable[[str], Optional[str]]] = None,
        **kwargs,
    ):
        super().__init__(action=action, **kwargs)
        self._categories = categories
        self._classifier = classifier

        self._patterns: Dict[str, List[re.Pattern]] = {}
        for cat, pats in _DEFAULT_CATEGORIES.items():
            if categories is None or cat in categories:
                self._patterns[cat] = pats

        if custom_categories:
            for cat, raw_pats in custom_categories.items():
                compiled = [re.compile(p, re.IGNORECASE) for p in raw_pats]
                self._patterns.setdefault(cat, []).extend(compiled)

    def evaluate(self, text: str) -> GuardResult:
        detected: List[str] = []

        for cat, patterns in self._patterns.items():
            for pat in patterns:
                if pat.search(text):
                    detected.append(cat)
                    break

        if not detected and self._classifier is not None:
            result = self._classifier(text)
            if result:
                detected.append(result)

        if detected:
            classification = ", ".join(sorted(detected))
            return GuardResult(
                action=self._action,
                score=round(min(1.0, len(detected) * 0.3 + 0.4), 2),
                guard_name=self.name,
                classification=classification,
                explanation=f"Sensitive topic(s) detected: {classification}",
            )

        return GuardResult(guard_name=self.name)

File: sdk/typescript/src/guard/sensitive-topic.ts
/**
 * Sensitive topic detection guard.
 *
 * Uses keyword/regex dictionaries for fast-path detection of sensitive
 * content categories. An optional user-provided classifier handles
 * ambiguous cases.
 *
 * Patterns must stay in sync with: sdk/python/src/openlit/guard/sensitive_topic.py
 */

import { Guard, GuardPhase, GuardResult, makeGuardResult, GuardOptions } from './base';

const DEFAULT_CATEGORIES: Record<string, RegExp[]> = {
  violence: [
    /\b(?:kill|murder|assault|attack|weapon|bomb|shoot|stab|torture|terrorism|massacre|genocide)\b/i,
  ],
  politics: [
    /\b(?:democrat|republican|election\s+fraud|political\s+party|vote\s+rigging|coup|insurrection)\b/i,
  ],
  substance_use: [
    /\b(?:cocaine|heroin|methamphetamine|fentanyl|drug\s+(?:deal|traffick)|overdose|illegal\s+drugs)\b/i,
  ],
  mental_health: [
    /\b(?:suicid(?:e|al)|self[- ]harm|eating\s+disorder|anorexia|bulimia)\b/i,
  ],
  discrimination: [
    /\b(?:racial\s+slur|white\s+supremac|ethnic\s+cleansing|hate\s+(?:speech|crime))\b/i,
  ],
  adult_content: [
    /\b(?:pornograph|explicit\s+sexual|nude\s+images|sex\s+trafficking)\b/i,
  ],
};

export interface SensitiveTopicOptions extends GuardOptions {
  categories?: Set<string> | string[];
  customCategories?: Record<string, string[]>;
  classifier?: (text: string) => string | null;
}

export class SensitiveTopic extends Guard {
  readonly name = 'sensitive_topic';
  readonly phases = [GuardPhase.PREFLIGHT, GuardPhase.POSTFLIGHT];

  private readonly _patterns: Record<string, RegExp[]>;
  private readonly _classifier?: (text: string) => string | null;

  constructor(opts: SensitiveTopicOptions = {}) {
    super({ action: opts.action ?? 'warn', maxScanLength: opts.maxScanLength });
    this._classifier = opts.classifier;

    const allowedCategories = opts.categories
      ? new Set(opts.categories)
      : null;

    this._patterns = {};
    for (const [cat, pats] of Object.entries(DEFAULT_CATEGORIES)) {
      if (allowedCategories === null || allowedCategories.has(cat)) {
        this._patterns[cat] = [...pats];
      }
    }

    if (opts.customCategories) {
      for (const [cat, rawPats] of Object.entries(opts.customCategories)) {
        const compiled = rawPats.map((p) => new RegExp(p, 'i'));
        if (!this._patterns[cat]) {
          this._patterns[cat] = [];
        }
        this._patterns[cat].push(...compiled);
      }
    }
  }

  evaluate(text: string): GuardResult {
    const detected: string[] = [];

    for (const [cat, patterns] of Object.entries(this._patterns)) {
      for (const pat of patterns) {
        if (pat.test(text)) {
          detected.push(cat);
          break;
        }
      }
    }

    if (detected.length === 0 && this._classifier) {
      const result = this._classifier(text);
      if (result) {
        detected.push(result);
      }
    }

    if (detected.length > 0) {
      const classification = [...detected].sort().join(', ');
      return makeGuardResult({
        action: this._action,
        score: Math.round(Math.min(1.0, detected.length * 0.3 + 0.4) * 100) / 100,
        guardName: this.name,
        classification,
        explanation: `Sensitive topic(s) detected: ${classification}`,
      });
    }

    return makeGuardResult({ guardName: this.name });
  }
}
